pychunkedgraph/graph/edges/utils.py

```diff
@@ -86,35 +87,62 @@ def merge_cross_edge_dicts_multiple(x_edges_d1: Dict, x_edges_d2: Dict) -> Dict:
 
 
 def categorize_edges(
-    meta: ChunkedGraphMeta, node_ids: np.ndarray, edges: Edges
-) -> Tuple[Edges, Edges]:
+    meta: ChunkedGraphMeta, supervoxels: np.ndarray, edges: Edges
+) -> Tuple[Edges, Edges, Edges]:
     """
     Find edges and categorize them into:
     `in_edges`
-        between given node_ids
-        (sv1, sv2) - sv1 in node_ids and sv2 in node_ids
+        between given supervoxels
+        (sv1, sv2) - sv1 in supervoxels and sv2 in supervoxels
     `out_edges`
-        originating from given node_ids but within chunk
-        (sv1, sv2) - sv1 in node_ids and sv2 not in node_ids
+        originating from given supervoxels but within chunk
+        (sv1, sv2) - sv1 in supervoxels and sv2 not in supervoxels
     `cross_edges`
-        originating from given node_ids but crossing chunk boundary
+        originating from given supervoxels but crossing chunk boundary
     """
-    mask1 = np.in1d(edges.node_ids1, node_ids)
-    mask2 = np.in1d(edges.node_ids2, node_ids)
+    from ..utils.context_managers import TimeIt
+
+    mask1 = np.isin(edges.node_ids1, supervoxels)
+    mask2 = np.isin(edges.node_ids2, supervoxels)
     in_mask = mask1 & mask2
     out_mask = mask1 & ~mask2
 
+    print("np.sum(in_mask)", np.sum(in_mask))
+
     in_edges = edges[in_mask]
     all_out_edges = edges[out_mask]  # out_edges + cross_edges
+
     edge_layers = get_cross_chunk_edges_layer(meta, all_out_edges.get_pairs())
     cross_edges_mask = edge_layers > 1
     out_edges = all_out_edges[~cross_edges_mask]
     cross_edges = all_out_edges[cross_edges_mask]
     return (in_edges, out_edges, cross_edges)
 
 
+def categorize_edges_v2(
+    meta: ChunkedGraphMeta,
+    supervoxels: np.ndarray,
+    edges: Edges,
+    l2id_children_d: Dict,
+    get_sv_parents: Callable,
+) -> Tuple[Edges, Edges, Edges]:
+    """ Faster version of categorize_edges(), avoids looping over L2 IDs. """
+    node_ids1 = get_sv_parents(edges.node_ids1)
+    node_ids2 = get_sv_parents(edges.node_ids2)
+
+    layer_mask1 = chunk_utils.get_chunk_layers(meta, node_ids1) > 1
```

Comment:
Why do you need this mask? Shouldn't `node_ids1` always have layer > 1 as they are the parents of supervoxels?

Reply:
No the vectorized function `get_sv_parents` returns the supervoxel if a parent is not found in this case, these are along `[:,1]` (outside a component) so it is expected that we do not have that info/don't need it to categorize edges.